Pass the certificate as a parameter to the import_dashboards script

monicasarbu opened the pull request:
Add `-ca` and `-key` option to the `import_dashboards` script in order to be able to pass the certificate. 
This PR is a workaround for now, until `import_dashboards` will be part of the Beat executable.

andrewkroh commented:
I was thinking we should have an option for a client cert and key too. This would cover the case where shield is setup to require TLS mutual auth.

monicasarbu commented:
@andrewkroh Yes, you are right. I added another option `-key` to be able to pass the the certificate key. I have updated also the description of the PR.

andrewkroh commented on libbeat/dashboards/import_dashboards.go:
```diff
@@ -91,6 +95,8 @@ func DefineCommandLine() (*CommandLine, error) {
 	cl.flagSet.BoolVar(&cl.opt.OnlyDashboards, "only-dashboards", false, "Import only dashboards together with visualizations and searches. By default import both, dashboards and the index-pattern.")
 	cl.flagSet.BoolVar(&cl.opt.OnlyIndex, "only-index", false, "Import only the index-pattern. By default imports both, dashboards and the index pattern.")
 	cl.flagSet.BoolVar(&cl.opt.Snapshot, "snapshot", false, "Import dashboards from snapshot builds.")
+	cl.flagSet.StringVar(&cl.opt.Certificate, "ca", "", "Certificate for SSL client authentication.")
```
There should be three options like `cacert`, `cert`, and `key`. I chose those names because they are the same as `curl`. And curl also allows a key passphrase to be specified with the `pass` CLI flag.

andrewkroh replied:
The description of the `cacert` param should say that the file may contain multiple PEM encoded CA certificates.

monicasarbu replied:
@andrewkroh I was working in the meantime on a solution where you can pass the `-cert` the client certificate, `-certkey` certificate key, and `-ca` certificate authority. I will change the code to use the proposed names as `curl` does.

andrewkroh commented on libbeat/dashboards/import_dashboards.go:
```diff
@@ -133,11 +147,27 @@ func New() (*Importer, error) {
+		tls, err = outputs.LoadTLSConfig(&outputs.TLSConfig{
+			Certificate: outputs.CertificateConfig{
+				Certificate: cl.opt.Certificate,
+				Key:         cl.opt.Key,
```
You could also set `Passphrase: cl.opt.Pass` to offer a way to specify a password for an encrypted key file.

monicasarbu replied:
@andrewkroh Yes true, but as it's a temporary solution, I think it is better to minimize the changes.

andrewkroh commented on libbeat/dashboards/import_dashboards.go:
```diff
@@ -107,6 +113,14 @@ func (cl *CommandLine) ParseCommandLine() error {
+	}
+
+	if cl.opt.Key != "" && cl.opt.Certificate == "" {
+		return errors.New("ERROR: A certificate needs to be passed as well by using the -ca option.")
```
`s/-ca/-cert/`

monicasarbu replied:
done

andrewkroh commented on libbeat/dashboards/import_dashboards.go:
```diff
@@ -91,6 +95,8 @@ func DefineCommandLine() (*CommandLine, error) {
 	cl.flagSet.BoolVar(&cl.opt.OnlyIndex, "only-index", false, "Import only the index-pattern. By default imports both, dashboards and the index pattern.")
 	cl.flagSet.BoolVar(&cl.opt.Snapshot, "snapshot", false, "Import dashboards from snapshot builds.")
+	cl.flagSet.StringVar(&cl.opt.Certificate, "ca", "", "Certificate for SSL client authentication.")
+	cl.flagSet.StringVar(&cl.opt.Key, "key", "", "Client Certificate Key")
```
The `cert` and `key` descriptions should mention PEM. (We don't support other formats.)

monicasarbu replied:
done

monicasarbu commented:
@andrewkroh I addressed your comments, so please have another look.
